Step2 리뷰 요청드립니다.

deokmoon opened the pull request:
안녕하세요 리뷰어님

step1 때 피드백 주신 내용 반영 및 연관관계 매핑에 대해 고민하며 미션 진행해봤습니다.

부족함이 많지만, 리뷰 해주시면 고민해보고 보완하도록 하겠습니다!
이번에도 잘 부탁드리겠습니다.

감사합니다!

deokmoon committed:
Refactor: properties -> yml 파일로 변경

deokmoon committed:
Feat: Hands-on Part1 머지

deokmoon committed:
Docs: readme 요구사항 작성

deokmoon committed:
Fix: yml 추가 적용(h2 console 및 mysql 방식)

deokmoon committed:
Fix: yml datasource url 수정

deokmoon committed:
Docs: 공통 timestamp Entity 관련 내용 작성

deokmoon committed:
Refactor: repository 관련 소스 repository패키지로 이동

deokmoon committed:
Feat: 공통 timestamp 적용할 BaseDateEntity 클래스 생성

deokmoon committed:
Feat: 요구사항에 맞게 엔티티의 각 필드 구현 (getter 제외)

deokmoon committed:
Test: Answer Entity 테스트코드 작성

deokmoon committed:
Test: Question Entity 테스트코드 작성(null 검증)

deokmoon committed:
Feat: Question Entity is not null 컬럼 검증 적용

deokmoon committed:
Test: User Entity is not null 컬럼 테스트코드 작성

deokmoon committed:
Feat: User Entity is not null 검증 적용, user_id unique 적용

deokmoon committed:
fix: yml 수정

deokmoon committed:
fix(domain): questionid not null 제거

deokmoon committed:
feat(main): EnableJpaAuditing 어노테이션 적용

deokmoon committed:
test(repository): AnswerRepositoryTest 테스트코드 작성

deokmoon committed:
test(repository): DeleteHistory 엔티티관련 테스트코드 작성

deokmoon committed:
feat(domain): DeleteHistory getter(id) 적용

deokmoon committed:
refactor(repository): 불필요한 import 삭제

deokmoon committed:
test(repository): Question 엔티티의 repository 테스트코드 작성

deokmoon committed:
test(repository): User 엔티티의 repository 테스트코드 작성

deokmoon committed:
docs: test 코드 및 yml 최신화

deokmoon committed:
refactor: readme dir 변경

deokmoon committed:
refactor: application.yml h2 방식 embedded mode로 재설정

deokmoon committed:
docs: application.yml 최신화

deokmoon committed:
fix(yml): ddl 전략  create 로 수정

deokmoon committed:
docs: 연관 관계 매핑 요구사항 작성

deokmoon committed:
refactor(domain): step1 피드백 반영
 - 생략 가능한 @table 어노테이션 삭제

deokmoon committed:
refactor(domain): step1 피드백 반영
 - "" 검증, isEmpty()로 대체

deokmoon committed:
fix(question,answer): 연관관계 매핑 설정 및 이에 따른 소스코드 수정

deokmoon committed:
fix(user,answer): 연관관계 매핑 설정 및 이에 따른 소스코드 수정 및 getter/setter 추가

deokmoon committed:
feat(user,answer, question): 연관관계 매핑 설정 및 이에 따른 소스코드 수정

deokmoon committed:
fix: 연관관계 매핑 설정에 따른 기존 데이터 중심의 설계 소스 수정

deokmoon committed:
feat(domain): 영속성전이(persist), 지연로딩 설정

deokmoon committed:
test(fixture): 테스트 픽스터 작성

deokmoon committed:
feat: answer 연관관계 npe 방지를 위한 addQuestion 구현

deokmoon committed:
test: answer 저장 및 cascade 테스트

deokmoon committed:
refactor: test fixture로 인한 리팩토링

deokmoon committed:
refactor: test fixture로 인한 리팩토링으로 삭제

deokmoon committed:
refactor: toString 및 불필요한 setter 삭제

deokmoon committed:
feat(yml): create-drop으로 재수정

deokmoon committed:
refactor: test fixture로 인한 수정

deokmoon committed:
test: 삭제이력 조회 테스트코드 작성

deokmoon commented:
혹시 머지 시 충돌 발생할까요 ..?

deokmoon committed:
fix: git 충돌 해결을 위한 push

deokmoon committed:
fix: git 충돌 해결을 위한 push

deokmoon committed:
fix: git 충돌 해결 중 삭제파일 복구로 인한 재삭제 처리

geunhwanlee commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -17,17 +21,20 @@
     @Column(nullable = false)
     private boolean deleted = false;
-    private Long writerId;
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
```
cascadeType에 PERSIST를 넣으면 어떻게 동작할까요? 🤔
그리고 PERSIST를 넣은 이유가 궁금합니다! 🙏

geunhwanlee commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -17,17 +21,20 @@
     private boolean deleted = false;
-    private Long writerId;
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
+    @JoinColumn(name = "QUESTION_ID")
+    private Question question;
```
`@JoinColumn`을 생략하면 어떻게 동작할까요?

deokmoon replied:
default 전략으로 "[필드명]_[참조하는 테이블의 기본키 컬럼명]" 으로 설정됩니다.
![joincolumn](https://user-images.githubusercontent.com/47620373/201684748-18764d5f-5ea6-4795-bafd-c248325ea7e6.JPG)

명시하는 것이 좋다고 생각하여, default로 사용했었습니다. 무조건 생각했던 내용도 덕분에 한번 더 찾아보게 되었습니다. 혹시 실무에선 joincolumn 어노테이션을 생략하는 전략으로 많이 사용하나요?

geunhwanlee replied:
정답은 없고 상황마다 사용하는 전략도 다 다를거에요.

개인적으로는 가독성 측면에서 생략할수 있다면 생략하는걸 선호합니다. 😀

명시하는것이 좋다고 생각하는 이유가 있다면 명시하는게 맞습니다! 👍🏻

geunhwanlee commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -17,17 +21,20 @@
+    @JoinColumn(name = "QUESTION_ID")
+    private Question question;
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
+    @JoinColumn(name = "WRITER_ID")
+    private User user;
```
```suggestion
    private User writer;
```
`@JoinColumn`을 생략하고 변수명을 writer로 쓰면 어떻게 동작할까요?

deokmoon replied:
현재 양방향참조관계를 설정하여, @JoinColumn을 생략하고 변수명을 writer로만 바꾸면 빌드 시 에러가 발생하게 됩니다. 연관관계 매핑 중 mappedBy 값 또한 writer로 변경해줘야지 원하는대로 동작하게 됩니다.

해당 부분은 리뷰어님이 지적 해주시기 전까지 크게 고민해보지 않은 사항인데, 리뷰어님 덕분에 한번 수행해보고 내부적으로 어떻게 처리되는지 확인하게 되었습니다.
감사합니다 👍

geunhwanlee commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -46,19 +53,21 @@ public Answer(Long id, User writer, Question question, String contents) {
 
-    public boolean isOwner(User writer) {
-        return this.writerId.equals(writer.getId());
+    public void addQuestion(Question question) {
+        this.question = question;
+        question.getAnswers().add(this);
```
`addQuestion` 메서드를 여러번 호출하면 하나의 `Question`에 같은 `Answer`가 여러개 연결될거같네요. 🤔

deokmoon replied:
앗... 그러네요 👍

geunhwanlee commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -99,12 +92,28 @@ public void setDeleted(boolean deleted) {
+        return question;
+    }
+
+    public void setQuestion(Question question) {
+        this.question = question;
+    }
```
도메인 모델에서 setter는 지양하는게 좋습니다. 😀

[참고링크](https://velog.io/@saint6839/도메인에서-Setter-사용을-지양하는-이유)

deokmoon replied:
해당 기능 addQuestion이 대체하면서 삭제하지 못했네요 
앞으로 리팩토링 시 더 주의해야겠습니다.

geunhwanlee commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -99,12 +92,28 @@ public void setDeleted(boolean deleted) {
                 "id=" + id +
-                ", writerId=" + writerId +
-                ", questionId=" + questionId +
+                ", writerId=" + user +
+                ", questionId=" + question +
```
```suggestion
                ", writerId=" + user.getId() +
                ", questionId=" + question.getId() +
```
참고로 양방향 참조관계에서 객체를 그대로 출력하게되면
서로가 서로의 `toString` 메서드를 호출하게 되어 스택오버플로우가 발생하게됩니다.

deokmoon replied:
앗.. 영한님 강의 들으면서 명심해야지 라고 생각했던 부분을 바로 실수했네요.
리뷰 감사드립니다 👍

geunhwanlee commented on src/main/java/qna/domain/User.java:
```diff
@@ -34,6 +36,12 @@ public class User extends BaseDateEntity{
     private String password;
     @Column(nullable = false, length = USER_ID_LENGTH, unique = true)
     private String userId;
+    @OneToMany(mappedBy = "user")
+    private List<Answer> answers = new ArrayList<Answer>();
+    @OneToMany(mappedBy = "user")
+    private List<Question> questions = new ArrayList<Question>();
+    @OneToMany(mappedBy = "user")
+    private List<DeleteHistory> deleteHistories = new ArrayList<DeleteHistory>();
```
양방향 연관관계 설정을 잘 해주셨네요!

다만 `User`의 경우 양방향 연관관계가 필요한지 의문이 들어요. 🤔
`User`에서 `질문`, `답변`, `삭제이력`과의 객체참조를 만들어야할까요?

deokmoon replied:
사실 현재 요구사항에선, User인 경우 양방향연관관계가 필요 없이, 각 도메인(question, answer, deleteHistories)에서 User로 단방향 연관관계만 있어도 된다고 생각합니다.

'다만 서비스 확장에 따른 사용자의 내글 보기 등의 기능이 생기는 경우, 양방향 연관관계가 필요하지 않을까!' 라고 생각하며 양방향 연관관계를 설정하였습니다.

geunhwanlee commented on src/test/java/qna/domain/AnswerTest.java:
```diff
@@ -1,25 +1,45 @@
     void writer_is_not_null_test() {
-        assertThatThrownBy(() -> new Answer(null, QuestionTest.Q1, "content"))
+        // given
+        User writer = TestUserFactory.create("JAVAJIGI");
```
테스트용 팩토리 클래스의 역할이 모호한것 같아요!

테스트용 팩토리 메서드를 통해 테스트 객체를 만들기보다는
도메인 객체의 생성자나 팩토리 메서드를 통해 만드는건 어떨까요? 🤔

deokmoon replied:
처음 생각한 설계는 test fixture를 위해 테스트용 팩토리 메서드를 생성한 의도하였습니다. 그리고 프로덕트 코드와 분리? 지저분하게 만들지 않겠다라는 생각 + 무작정 static을 피하려는 생각이었습니다.

근환님의 리뷰를 통해 어떤점이 모호할까? 부터 생각을 해보았고, 기존 테스트팩토리메서드에서 하는 역할이 그냥 생성자와 다를게 없나? 차라리 무분별한 생성을 막기 위해 팩토리메서드를 설정하는 것이 맞나? 라는 고민을 했습니다.  

고민 끝에 no args constructor는 protected를 하되, test 용, product 용 정적 팩토리 메서드를 적용하였습니다. 이유는 테스트용 팩토리 메서드를 썼을 떄의 장점이 딱히 없어서입니다. 

제가 얻은 결론이 근환님이 의도하신 내용이 맞는지 한번 여쭤보고싶습니다. 참고할만한 내용도 주시면 감사하겠습니다!

geunhwanlee reviewed (changes requested):
2단계 미션 잘 해주셨어요! 👍🏻

다만 몇가지 아쉬운 부분이 있어서 코멘트 드렸어요.

확인해보시고 재 리뷰요청 주세요. 🙏

deokmoon committed:
fix(domain): import 정리 및 toString 참조객체 수정
 - user -> user.getUserId

deokmoon committed:
fix(domain): addQeustion 메서드 수정 및 user 참조객체 필드명 변경
 - user -> writer로 변수명 직관적으로 수정
 - addQuestion 버그 수정
 - question setter 삭제
 - toString() 메서드 user -> user.getUserId

deokmoon committed:
refactor(domain): user 참조객체 필드명 변경에 따른 호출 메서드명 수정
 - user -> writer로 변수명 직관적으로 수정

deokmoon committed:
refactor(domain): user 참조객체 필드명 변경에 따른 호출 메서드명 수정
 - user -> writer로 변수명 직관적으로 수정

deokmoon committed:
refactor(domain): user 참조객체 필드명 변경에 따른 mappedBy user -> writer 수정

deokmoon committed:
refactor: 테스트 팩토리 메서드 도메인 객체의 팩토리 메서드 활용하도록 수정

deokmoon committed:
refactor: 각 도메인 객체 팩토리 메서드 적용에 따른 수정

geunhwanlee commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -3,39 +3,49 @@
 @Entity
-@Table(name = "answer")
 public class Answer extends BaseDateEntity{
+    // test 용
+    public static Answer create(User writer, Question question) {
+        return new Answer(writer, question, "contents");
+    }
```
제가 드린 코멘트에 오해가 있었네요. 😂

테스트용 팩토리 메서드를 만드는것보다
도메인 객체의 생성자 또는 팩토리 메서드를 사용해서 생성하는건 어떨까 하는게 제 의도였습니다. 🙏

테스트는 테스트일뿐 도메인 객체가 테스트용 메서드를 갖는다는건 좋지 않다고 생각해요.

테스트용 픽스처를 만드실때

> `Answer.create(writer, question, "contents");`
> 또는
> `new Answer(writer, question, "contents");`

이렇게 써도 괜찮지 않을까 싶어서 코멘트 드렸던거에요. 😀

물론 테스트 팩토리 메서드를 만들어서 테스트에 불필요한 값들을 임의로 지정해주는것도 테스트의 가독성 측면에서 의미가 있습니다. 👍🏻

테스트 픽스처와 관련해서 이런 라이브러리도 있네요!
이번 미션에서 적용하실수는 없지만 한번 참고해보셔도 좋을것같아요. 🐵
[Fixture Monkey](https://leeheefull.tistory.com/m/27)

geunhwanlee reviewed (approved):
리뷰내용 잘 반영해주셨어요! 👍🏻

다만 테스트 픽스처 생성 관련해서는 제 의도와 다르게 이해하신거같아요. 😭
이부분은 다음단계때 수정 부탁드립니다. 🙏
(원래 하셨던 방향도 좋습니다. 👍🏻)

그럼 이번 단계는 여기서 머지하도록 할게요.
